[VPlan] Compute induction end values in VPlan.

fhahn opened the pull request:
Use createDerivedIV to compute IV end values directly in VPlan, instead
of creating them up-front.

This allows updating IV users outside the loop as follow-up.

Depends on https://github.com/llvm/llvm-project/pull/110004 and
https://github.com/llvm/llvm-project/pull/109975.

fhahn committed:
[VPlan] Compute induction end values in VPlan.

Use createDerivedIV to compute IV end values directly in VPlan, instead
of creating them up-front.

This allows updating IV users outside the loop as follow-up.

Depends on https://github.com/llvm/llvm-project/pull/110004 and
https://github.com/llvm/llvm-project/pull/109975.

fhahn commented:
Now that https://github.com/llvm/llvm-project/pull/110577 landed, this should be ready for review.

Some VPlan tests still need updating, but the relevant IR tests have been updated already. It still needs some updates for outer loop vectorization.

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8858,13 +8794,64 @@ static void addCanonicalIVRecipes(VPlan &Plan, Type *IdxTy, bool HasNUW,
+  VPValue *Step;
+  Type *ScalarTy;
+  bool IsCanonical = false;
+  if (auto *WideIV = dyn_cast<VPWidenIntOrFpInductionRecipe>(PhiR)) {
```
Seems like a common VPWidenInductionRecipe base class would be useful, covering Int or FP or Pointer?

fhahn replied:
Put up  https://github.com/llvm/llvm-project/pull/120008 to do so. Also included the changes here

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8858,13 +8794,64 @@ static void addCanonicalIVRecipes(VPlan &Plan, Type *IdxTy, bool HasNUW,
+                                            VPValue *VectorTC) {
+  PHINode *OrigPhi;
+  const InductionDescriptor *ID;
+  VPValue *Start;
```
```suggestion
  VPValue *Start = PhiR.getStartValue();
```

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8858,13 +8794,64 @@ static void addCanonicalIVRecipes(VPlan &Plan, Type *IdxTy, bool HasNUW,
+        VectorTC, Step);
+  }
+
+  if (ScalarTy != TypeInfo.inferScalarType(EndValue)) {
+    EndValue = Builder.createScalarCast(Instruction::Trunc, EndValue, ScalarTy);
+  }
```
```suggestion
  // Worth explaining why these types may differ, and if so why trunc rather than zext/sext is needed.
  if (ScalarTy != TypeInfo.inferScalarType(EndValue))
    EndValue = Builder.createScalarCast(Instruction::Trunc, EndValue, ScalarTy);
```

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8858,13 +8794,64 @@ static void addCanonicalIVRecipes(VPlan &Plan, Type *IdxTy, bool HasNUW,
 }
 
+static VPValue *addResumeValuesForInduction(VPHeaderPHIRecipe *PhiR,
+                                            VPBuilder &Builder,
```
Better rename `Builder` according to its position, consistent with `ScalarPHBuilder`.

ayalz replied:
```suggestion
                                            VPBuilder &VectorPHBuilder,
```
consistent with `ScalarPHBuilder`.

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8858,13 +8794,64 @@ static void addCanonicalIVRecipes(VPlan &Plan, Type *IdxTy, bool HasNUW,
+  return ResumePhiRecipe;
+}
+
 /// Create resume phis in the scalar preheader for first-order recurrences and
 /// reductions and update the VPIRInstructions wrapping the original phis in the
```
```suggestion
/// Create resume phis in the scalar preheader for first-order recurrences, reductions and inductions, and update the VPIRInstructions wrapping the original phis in the
```

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8874,6 +8861,13 @@ static void addScalarResumePhis(VPRecipeBuilder &Builder, VPlan &Plan) {
+      ScalarPhiIRI->addOperand(ResumePhi);
+      continue;
+    }
     if (!isa<VPFirstOrderRecurrencePHIRecipe, VPReductionPHIRecipe>(VectorPhiR))
```
This case refers to truncating inductions?

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9635,7 +9629,6 @@ void VPDerivedIVRecipe::execute(VPTransformState &State) {
       State.Builder, CanonicalIV, getStartValue()->getLiveInIRValue(), Step,
       Kind, cast_if_present<BinaryOperator>(FPBinOp));
   DerivedIV->setName(Name);
-  assert(DerivedIV != CanonicalIV && "IV didn't need transforming?");
```
Such redundant recipes should be cleaned up, if generated at all?

fhahn replied:
Cleaned up in the latest version, except for the case where the vector trip count is constant null, which will only be set in prepareToExecute.

fhahn replied:
Yep

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10299,6 +10292,39 @@ bool LoopVectorizePass::processLoop(Loop *L) {
                                            EPI, &LVL, &CM, BFI, PSI, Checks,
                                            *BestMainPlan);
 
+        // Collect PHI nodes of wide inductions in the VPlan for the epilogue.
+        // Those will need their resume-values computed from the main vector
+        // loop. Others can be removed in the main VPlan.
+        SmallPtrSet<PHINode *, 2> WidenedPhis;
+        for (VPRecipeBase &R :
+             BestEpiPlan.getVectorLoopRegion()->getEntryBasicBlock()->phis()) {
+          if (!isa<VPWidenIntOrFpInductionRecipe,
+                   VPWidenPointerInductionRecipe>(&R))
+            continue;
+          if (isa<VPWidenIntOrFpInductionRecipe>(&R))
+            WidenedPhis.insert(
+                cast<VPWidenIntOrFpInductionRecipe>(&R)->getPHINode());
+          else
+            WidenedPhis.insert(
+                cast<PHINode>(R.getVPSingleValue()->getUnderlyingValue()));
+        }
+        for (VPRecipeBase &R :
+             *cast<VPIRBasicBlock>(BestMainPlan->getScalarHeader())) {
+          auto *VPIRInst = cast<VPIRInstruction>(&R);
+          auto *IRI = dyn_cast<PHINode>(&VPIRInst->getInstruction());
+          if (!IRI)
+            break;
+          if (WidenedPhis.contains(IRI) ||
+              !LVL.getInductionVars().contains(IRI))
+            continue;
+          VPRecipeBase *ResumePhi =
+              VPIRInst->getOperand(0)->getDefiningRecipe();
+          VPIRInst->setOperand(0, BestMainPlan->getOrAddLiveIn(
+                                      Constant::getNullValue(IRI->getType())));
+          ResumePhi->eraseFromParent();
+        }
+        VPlanTransforms::removeDeadRecipes(*BestMainPlan);
+
```
Can this be outlined and/or placed elsewhere, too detailed for processLoop() which is already too long and doing too much.

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10299,6 +10292,39 @@ bool LoopVectorizePass::processLoop(Loop *L) {
 
+        // Collect PHI nodes of wide inductions in the VPlan for the epilogue.
+        // Those will need their resume-values computed from the main vector
+        // loop. Others can be removed in the main VPlan.
```
removed?

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10299,6 +10292,39 @@ bool LoopVectorizePass::processLoop(Loop *L) {
+          if (WidenedPhis.contains(IRI) ||
+              !LVL.getInductionVars().contains(IRI))
+            continue;
+          VPRecipeBase *ResumePhi =
+              VPIRInst->getOperand(0)->getDefiningRecipe();
+          VPIRInst->setOperand(0, BestMainPlan->getOrAddLiveIn(
+                                      Constant::getNullValue(IRI->getType())));
+          ResumePhi->eraseFromParent();
```
Clean up redundant resume Phi recipes created for inductions that were not widened? VPlan's DCE cannot collect them because of cross-VPlan def-use?

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10299,6 +10292,39 @@ bool LoopVectorizePass::processLoop(Loop *L) {
+            continue;
+          VPRecipeBase *ResumePhi =
+              VPIRInst->getOperand(0)->getDefiningRecipe();
+          VPIRInst->setOperand(0, BestMainPlan->getOrAddLiveIn(
+                                      Constant::getNullValue(IRI->getType())));
```
Deserves some explanation/comment.

ayalz reviewed (commented):
Nice step leveraging ResumePhi and DerivedIV recipes!
Patch should be marked NFC?

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn committed:
!fixup address first set of comments, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn committed:
!fixup address latest comments, thanks!

fhahn committed:
!fixup simplify derived IV

fhahn reviewed (commented):
Addressed latest comments and updated tests

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2641,33 +2628,21 @@ void InnerLoopVectorizer::createVectorLoopSkeleton(StringRef Prefix) {
-
-    // Compute the end value for the additional bypass (if applicable).
+    // Compute the end value for the additional bypass.
     if (MainVectorTripCount) {
```
```suggestion
```
?

fhahn replied:
Done thanks

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2641,33 +2628,21 @@ void InnerLoopVectorizer::createVectorLoopSkeleton(StringRef Prefix) {
-                                    Step, II.getKind(), II.getInductionBinOp());
-    EndValue->setName("ind.end");
-
-    // Compute the end value for the additional bypass (if applicable).
```
"(if applicable)" should be removed?

fhahn replied:
Yep, we now always compute the value he

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7827,8 +7759,9 @@ DenseMap<const SCEV *, Value *> LoopVectorizationPlanner::executePlan(
-                             ILV.getOrCreateVectorTripCount(nullptr), State);
+  BestVPlan.prepareToExecute(
+      ILV.getTripCount(),
+      ILV.getOrCreateVectorTripCount(ILV.LoopVectorPreHeader), State);
```
How/Is this change related?

fhahn replied:
Previously getOrCreateVectorTripCount would be called by createInductionResumeVPValues, which has now been removed, so we need to pass `ILV.LoopVectorPreHeader` here create the vector trip count IR

ayalz replied:
Ahh, previously vector trip count was created earlier, so here it was effectively only getting it, so null could be passed.

Can commit this independently, removing the dependence that this not be the first time it is called.

fhahn replied:
Done in f9120dc2a60aedcab5ce99e40b6a2bd3849f0bb9, thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9717,13 +9720,13 @@ void VPDerivedIVRecipe::execute(VPTransformState &State) {
   DerivedIV->setName(Name);
-  assert(DerivedIV != CanonicalIV && "IV didn't need transforming?");
-
+  // Index may only bet set to constant 0 in prepareToExecute.
```
```suggestion
  // Index may only be set to constant 0 in prepareToExecute.
```

fhahn replied:
Fixed thanks

fhahn committed:
!fixup address latest comment, thanks

fhahn committed:
!fixup update remaining tests

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn commented:
> Patch should be marked NFC?

Yep it should be NFC modulo slight reordering of phis/induction-end computations

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn committed:
!fixup update last remaining phaseordering test

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn committed:
!fixup use VPWidenInductionRecipe.

fhahn committed:
[VPlan] Get types and step from VPWidenPointerInductionRecipe (NFC).

fhahn committed:
[VPlan] Add VPWidenInduction recipe as common base class (NFC).

This helps to simplify some existing code and new code
(https://github.com/llvm/llvm-project/pull/112145)

fhahn committed:
Merge branch 'vpwideninductionrecipe' into vplan-compute-iv-end-values

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn commented:
Updated after landing https://github.com/llvm/llvm-project/pull/120008

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

alexey-bataev commented on llvm/test/Transforms/LoopVectorize/AArch64/low_trip_count_predicates.ll:
```diff
@@ -74,6 +74,7 @@ define void @low_vf_ic_is_better(ptr nocapture noundef %p, i32 %tc, i16 noundef
 ; CHECK-VS1-NEXT:    [[N_VEC:%.*]] = sub i64 [[TMP3]], [[N_MOD_VF]]
 ; CHECK-VS1-NEXT:    [[TMP18:%.*]] = call i64 @llvm.vscale.i64()
 ; CHECK-VS1-NEXT:    [[TMP19:%.*]] = mul i64 [[TMP18]], 16
+; CHECK-VS1-NEXT:    [[TMP40:%.*]] = add i64 [[TMP0]], [[N_VEC]]
```
Unused instruction?

fhahn replied:
Yes unfortunately now we create the end VPValues during VPlan construction and now during the legacy epilogue skeleton code generation there's no convenient way to remove them for the epilogue only case I think

fhahn replied:
Added TODO, thanks!

fhahn replied:
Should be gone now, thanks

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9439,6 +9418,20 @@ VPlanPtr LoopVectorizationPlanner::buildVPlan(VFRange &Range) {
   bool HasNUW = true;
   addCanonicalIVRecipes(*Plan, Legal->getWidestInductionType(), HasNUW,
                         DebugLoc());
+
```
Would be good to reduce the disparity between native and legacy VPlan constructions (ultimately converging). Perhaps have the former also record Ingredient2Recipe for widened inductions it creates during VPInstructionsToVPRecipes(), to then use getRecipe(P) as above, rather than searching linearly for each as below?

fhahn replied:
Yep, adjusted, thanks!

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn committed:
!fixup update tests after update to main

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationPlanner.h:
```diff
@@ -233,8 +233,8 @@ class VPBuilder {
                                      FPMathOperator *FPBinOp, VPValue *Start,
-                                     VPCanonicalIVPHIRecipe *CanonicalIV,
-                                     VPValue *Step, const Twine &Name = "") {
+                                     VPValue *CanonicalIV, VPValue *Step,
```
Added a comment, thanks.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2733,46 +2696,16 @@ static void addFullyUnrolledInstructionsToIgnore(
+  for (const auto &InductionEntry : Legal->getInductionVars()) {
+    PHINode *OrigPhi = InductionEntry.first;
+    const InductionDescriptor &II = InductionEntry.second;
+    createInductionAdditionalBypassValue(
```
Done, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8919,13 +8836,63 @@ static void addCanonicalIVRecipes(VPlan &Plan, Type *IdxTy, bool HasNUW,
+  const InductionDescriptor &ID = WideIV->getInductionDescriptor();
+  Type *ScalarTy = TypeInfo.inferScalarType(WideIV);
+  bool IsCanonical = false;
+  if (auto *WideIV = dyn_cast<VPWidenIntOrFpInductionRecipe>(PhiR)) {
```
Done, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8919,13 +8836,63 @@ static void addCanonicalIVRecipes(VPlan &Plan, Type *IdxTy, bool HasNUW,
+  bool IsCanonical = false;
+  if (auto *WideIV = dyn_cast<VPWidenIntOrFpInductionRecipe>(PhiR)) {
+    // Truncated wide inductions resume from the last lane of their vector value
+    // in the last vector iteration.
```
Done, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8920,13 +8836,63 @@ static void addCanonicalIVRecipes(VPlan &Plan, Type *IdxTy, bool HasNUW,
+  if (auto *WideIV = dyn_cast<VPWidenIntOrFpInductionRecipe>(PhiR)) {
+    // Truncated wide inductions resume from the last lane of their vector value
+    // in the last vector iteration.
+    if (ScalarTy != TypeInfo.inferScalarType(Step))
```
Done, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8920,13 +8836,63 @@ static void addCanonicalIVRecipes(VPlan &Plan, Type *IdxTy, bool HasNUW,
+        Start, VectorTC, Step);
+  }
+
+  // EndValue is based on the vector trip count (which has the same type as the
```
Done, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8919,13 +8836,63 @@ static void addCanonicalIVRecipes(VPlan &Plan, Type *IdxTy, bool HasNUW,
+  }
+
+  VPValue *EndValue = VectorTC;
+  if (!IsCanonical) {
```
I don't think it can be moved, we need to compute the end value for all cases except wide canonical IVs

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8920,13 +8836,63 @@ static void addCanonicalIVRecipes(VPlan &Plan, Type *IdxTy, bool HasNUW,
   VPBuilder ScalarPHBuilder(ScalarPH);
+  VPBuilder VectorPHBuilder(
+      cast<VPBasicBlock>(Plan.getVectorLoopRegion()->getSinglePredecessor()));
   VPBuilder MiddleBuilder(MiddleVPBB, MiddleVPBB->getFirstNonPhi());
```
Done, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8935,9 +8901,20 @@ static void addScalarResumePhis(VPRecipeBuilder &Builder, VPlan &Plan) {
       break;
-    auto *VectorPhiR = cast<VPHeaderPHIRecipe>(Builder.getRecipe(ScalarPhiI));
-    if (!isa<VPFirstOrderRecurrencePHIRecipe, VPReductionPHIRecipe>(VectorPhiR))
+    auto *VectorPhiR = GetHeaderPhiRecipe(ScalarPhiI);
```
Sounds like a great cleanup, will do separately.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8935,9 +8901,20 @@ static void addScalarResumePhis(VPRecipeBuilder &Builder, VPlan &Plan) {
+            &Plan.getVectorTripCount())) {
+      ScalarPhiIRI->addOperand(ResumePhi);
       continue;
+    }
```
Done, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8935,9 +8901,20 @@ static void addScalarResumePhis(VPRecipeBuilder &Builder, VPlan &Plan) {
       continue;
+    }
+    if (!isa<VPFirstOrderRecurrencePHIRecipe, VPReductionPHIRecipe>(
+            VectorPhiR)) {
```
done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8935,9 +8901,20 @@ static void addScalarResumePhis(VPRecipeBuilder &Builder, VPlan &Plan) {
+    if (!isa<VPFirstOrderRecurrencePHIRecipe, VPReductionPHIRecipe>(
+            VectorPhiR)) {
+      assert(cast<VPWidenIntOrFpInductionRecipe>(VectorPhiR)->getTruncInst() &&
+             "should only skip truncated wide inductions");
```
Added a TODO, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9718,13 +9711,14 @@ void VPDerivedIVRecipe::execute(VPTransformState &State) {
   DerivedIV->setName(Name);
-  assert(DerivedIV != CanonicalIV && "IV didn't need transforming?");
-
+  // Index may only be set to constant 0 in prepareToExecute.
```
no, we only create the expression for the vector trip count late and set it there.

ayalz replied:
I still seem to be missing something here. The fact that prepareToExecute() creates the vector trip count (typically to non zero?) implies that derived IVs that are based on it will miss the opportunity to be folded by simplifyRecipes(), because it's too late.

fhahn replied:
Yes that is the issue; VectorTripCountV is only created during skeleton creation and set as underlying value here, so we miss this opportunity to fold. I adjusted the assert to check if the index is the vector trip count instead and expanded the comment.

Added a TODO to remove this once we construct VTC in VPlan, and can use it for simplifications.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10034,6 +10028,63 @@ LoopVectorizePass::LoopVectorizePass(LoopVectorizeOptions Opts)
                               !EnableLoopVectorization) {}
 
+/// Prepare \p MainPlan for vectorizing the main vector loop during epilogue
+/// vectorization. Remove ResumePhis from \p MainPlan for inductions if they
```
Done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10034,6 +10028,63 @@ LoopVectorizePass::LoopVectorizePass(LoopVectorizeOptions Opts)
+      WidenedPhis.insert(cast<VPWidenIntOrFpInductionRecipe>(&R)->getPHINode());
+    else
+      WidenedPhis.insert(
+          cast<PHINode>(R.getVPSingleValue()->getUnderlyingValue()));
```
Done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10034,6 +10028,63 @@ LoopVectorizePass::LoopVectorizePass(LoopVectorizeOptions Opts)
+    auto *IRI = dyn_cast<PHINode>(&VPIRInst->getInstruction());
+    if (!IRI)
+      break;
+    if (WidenedPhis.contains(IRI) || !Inductions.contains(IRI))
```
Done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10034,6 +10028,63 @@ LoopVectorizePass::LoopVectorizePass(LoopVectorizeOptions Opts)
+  }
+
+  using namespace VPlanPatternMatch;
+  VPBasicBlock *ScalarPHVPBB = MainPlan.getScalarPreheader();
```
Done , thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10034,6 +10028,63 @@ LoopVectorizePass::LoopVectorizePass(LoopVectorizeOptions Opts)
+  VPBasicBlock *ScalarPHVPBB = MainPlan.getScalarPreheader();
+  VPValue *VectorTC = &MainPlan.getVectorTripCount();
+  // If there is no suitable resume value for the canonical induction in the
+  // epilogue loop, create it.
```
Done, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10034,6 +10028,63 @@ LoopVectorizePass::LoopVectorizePass(LoopVectorizeOptions Opts)
+      })) {
+    VPBuilder ScalarPHBuilder(ScalarPHVPBB, ScalarPHVPBB->begin());
+    // When vectorizing the epilogue, create a resume phi for the
+    // canonical IV if no suitable resume phi was already created.
```
Yep, dropped thanks

fhahn committed:
!fixup address latest comments, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn committed:
fixup address latest comments, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9481,7 +9445,9 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
         *Plan, *PSE.getSE(), OrigLoop, UncountableExitingBlock, RecipeBuilder);
   }
-  addScalarResumePhis(RecipeBuilder, *Plan);
+  addScalarResumePhis(*Plan, [&RecipeBuilder](PHINode *P) {
```
Done, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10206,6 +10187,57 @@ LoopVectorizePass::LoopVectorizePass(LoopVectorizeOptions Opts)
+    EpiWidenedPhis.insert(
+        cast<PHINode>(R.getVPSingleValue()->getUnderlyingValue()));
+  }
+  for (VPRecipeBase &R : make_early_inc_range(
```
Removed again thanks.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10034,6 +10028,63 @@ LoopVectorizePass::LoopVectorizePass(LoopVectorizeOptions Opts)
+    // can be removed. The resume values for the scalar loop will be created
+    // during execution of EpiPlan.
+    VPRecipeBase *ResumePhi = VPIRInst->getOperand(0)->getDefiningRecipe();
+    VPIRInst->setOperand(
+        0, MainPlan.getOrAddLiveIn(Constant::getNullValue(IRI->getType())));
```
(thought I responended but cannot see the response here): VPIRinst wraps the header phi in the scalar loop, so it is probaby better to keep it here.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10206,6 +10187,57 @@ LoopVectorizePass::LoopVectorizePass(LoopVectorizeOptions Opts)
+  // will need their resume-values computed in the main vector loop. Others
+  // can be removed from the main VPlan.
+  SmallPtrSet<PHINode *, 2> EpiWidenedPhis;
+  for (VPRecipeBase &R :
```
Left the concrete tpye for now, as the type isn't directly 'obvious' like via dyn_cast/cast,etc

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10206,6 +10187,57 @@ LoopVectorizePass::LoopVectorizePass(LoopVectorizeOptions Opts)
+  VPValue *VectorTC = &MainPlan.getVectorTripCount();
+  // If there is no suitable resume value for the canonical induction in the
+  // scalar (which will become vector) epilogue loop, create it.
+  if (none_of(*MainScalarPH, [VectorTC](VPRecipeBase &R) {
```
done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10206,6 +10187,57 @@ LoopVectorizePass::LoopVectorizePass(LoopVectorizeOptions Opts)
+    ScalarPHBuilder.createNaryOp(
+        VPInstruction::ResumePhi,
+        {VectorTC, MainPlan.getOrAddLiveIn(ConstantInt::get(
+                       MainPlan.getCanonicalIV()->getScalarType(), 0))},
```
Done thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1063,6 +1063,14 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
 
   if (match(&R, m_Not(m_Not(m_VPValue(A)))))
     return R.getVPSingleValue()->replaceAllUsesWith(A);
+
```
Yes those 2 cases are covered by the DerivedIVs introduced in this patch.

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/sve-tail-folding-forced.ll:
```diff
@@ -11,6 +11,7 @@ target triple = "aarch64-unknown-linux-gnu"
 ; VPLANS-LABEL: Checking a loop in 'simple_memset'
 ; VPLANS:      VPlan 'Initial VPlan for VF={vscale x 1,vscale x 2,vscale x 4},UF>=1' {
 ; VPLANS-NEXT: Live-in vp<[[VFxUF:%.+]]> = VF * UF
+; VPLANS-NEXT: Live-in vp<[[VTC:%.+]]> = vector-trip-count
```
I think this is due to the test not matching the scalar prehaeder. Should we extend the scope of what the test covers?

fhahn replied:
Dropped, thanks

fhahn commented on llvm/test/Transforms/LoopVectorize/PowerPC/exit-branch-cost.ll:
```diff
@@ -192,6 +192,8 @@ exit:
 }
 ;.
+; CHECK: attributes #[[ATTR0:[0-9]+]] = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
+;.
```
No, run the script with an unneeded argument, removed, thanks

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/strided-accesses.ll:
```diff
@@ -348,7 +347,7 @@ define void @single_stride_int_iv(ptr %p, i64 %stride) {
 ; NOSTRIDED:       scalar.ph:
 ; NOSTRIDED-NEXT:    [[BC_RESUME_VAL:%.*]] = phi i64 [ [[N_VEC]], [[MIDDLE_BLOCK]] ], [ 0, [[VECTOR_SCEVCHECK]] ], [ 0, [[ENTRY:%.*]] ]
-; NOSTRIDED-NEXT:    [[BC_RESUME_VAL1:%.*]] = phi i64 [ [[IND_END]], [[MIDDLE_BLOCK]] ], [ 0, [[VECTOR_SCEVCHECK]] ], [ 0, [[ENTRY]] ]
+; NOSTRIDED-NEXT:    [[BC_RESUME_VAL1:%.*]] = phi i64 [ [[N_VEC]], [[MIDDLE_BLOCK]] ], [ 0, [[VECTOR_SCEVCHECK]] ], [ 0, [[ENTRY]] ]
```
Yep

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn reviewed (commented):
Split off the changes to simplify DerivedIV to e1833e3a7e9de126d0a71d439bb862a8e20a5e4f and use it to simplify `createScalarIVSteps`

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10078,6 +10053,56 @@ LoopVectorizePass::LoopVectorizePass(LoopVectorizeOptions Opts)
+    // can be removed. The resume values for the scalar loop will be created
+    // during execution of EpiPlan.
+    VPRecipeBase *ResumePhi = VPIRInst->getOperand(0)->getDefiningRecipe();
+    VPIRInst->setOperand(
+        0, MainPlan.getOrAddLiveIn(Constant::getNullValue(IRI->getType())));
```
> VPIRinst wraps the header phi in the scalar loop, so it is probaby better to keep it here.

Is this operand needed, set artificially to zero (can convey a type... ;-), or should it best be removed, keeping VPIRInst only as a wrapper of the header phi?

fhahn replied:
At the moment it is needed, otherwise it would complicate VPIRInstruction::execute; but in the future we can disconnect the scalar header for the main vector loop, thus removing the need to update them there. (There is code elsewhere that use the scalar phi to get the resume value from the main vector loop)

fhahn replied:
Both can be removed in the latest version, updated, thanks!

Disconnection should be able after https://github.com/llvm/llvm-project/pull/120918

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2733,46 +2655,40 @@ static void addFullyUnrolledInstructionsToIgnore(
+      // Fast-math-flags propagate from the original induction instruction.
+      if (isa_and_nonnull<FPMathOperator>(II.getInductionBinOp()))
+        BypassBuilder.setFastMathFlags(
+            II.getInductionBinOp()->getFastMathFlags());
```
Done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8955,24 +8854,83 @@ static void addCanonicalIVRecipes(VPlan &Plan, Type *IdxTy, bool HasNUW,
-    if (!isa<VPFirstOrderRecurrencePHIRecipe, VPReductionPHIRecipe>(VectorPhiR))
+    if (isa<VPWidenInductionRecipe>(VectorPhiR)) {
+      if (VPValue *ResumePhi = addResumePhiRecipeForInduction(
+              VectorPhiR, VectorPHBuilder, ScalarPHBuilder, TypeInfo,
```
Done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9762,13 +9732,18 @@ void VPDerivedIVRecipe::execute(VPTransformState &State) {
-
+  // If index is the vector trip count, the concrete value will only be set in
+  // prepareToExecute, leading to missed simplifications, e.g. if it is 0.
+  // TODO: Remove the special case for the vector trip count once it is computed
```
This case is slightly counter-intuitive, VTC == 0 can only happen when the original BTC == -1, but should be handled correctly with checking the canonical IV increment

fhahn replied:
Yep, will look into that separately, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10078,6 +10053,56 @@ LoopVectorizePass::LoopVectorizePass(LoopVectorizeOptions Opts)
+    VPRecipeBase *ResumePhi = VPIRInst->getOperand(0)->getDefiningRecipe();
+    VPIRInst->setOperand(
+        0, MainPlan.getOrAddLiveIn(Constant::getNullValue(IRI->getType())));
+    ResumePhi->eraseFromParent();
```
Added a run here; this cleans up the redundant IV computations discussed in one of the tests

fhahn committed:
!fixup address latest comments, thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2733,46 +2655,41 @@ static void addFullyUnrolledInstructionsToIgnore(
+      auto *BinOp = II.getInductionBinOp();
+      // Fast-math-flags propagate from the original induction instruction.
+      if (isa_and_nonnull<FPMathOperator>(BinOp))
+        BypassBuilder.setFastMathFlags(
+            II.getInductionBinOp()->getFastMathFlags());
```
```suggestion
        BypassBuilder.setFastMathFlags(BinOp->getFastMathFlags());
```

fhahn replied:
DOne thanks

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8955,24 +8855,79 @@ static void addCanonicalIVRecipes(VPlan &Plan, Type *IdxTy, bool HasNUW,
-/// Create resume phis in the scalar preheader for first-order recurrences and
-/// reductions and update the VPIRInstructions wrapping the original phis in the
-/// scalar header.
+/// Create and return a ResumePhi for \p PhiR, if it is wide induction recipe.
```
```suggestion
/// Create and return a ResumePhi for \p WideIV, unless it is truncated.
```

fhahn replied:
Updated thanks!

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn committed:
!fixup address comments, thanks!

fhahn committed:
!fixup fix formatting

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10078,6 +10049,57 @@ LoopVectorizePass::LoopVectorizePass(LoopVectorizeOptions Opts)
+    // There is no corresponding wide induction in the epilogue plan that would
+    // need a resume value. Set the operand in VPIRInst to zero, so ResumePhi
+    // can be removed. The resume values for the scalar loop will be created
+    // during execution of EpiPlan.
```
> The resume values for the scalar loop will be created
    // during execution of EpiPlan.

All resume values, or only those free of wide users in EpiPlan?

fhahn replied:
Should be all resume values for the scalar loop, those free of wide users are not need to resume in the epilogue vector loop

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10078,6 +10049,57 @@ LoopVectorizePass::LoopVectorizePass(LoopVectorizeOptions Opts)
+    if (EpiWidenedPhis.contains(IRI))
+      continue;
+    // There is no corresponding wide induction in the epilogue plan that would
+    // need a resume value. Set the operand in VPIRInst to zero, so ResumePhi
```
Emphasize that this zero is artificial, along with a FIXME to remove it (one way or another)?

fhahn replied:
Both can be removed in the latest version, updated, thanks!

ayalz reviewed (approved):
LGTM, added a couple of final thoughts.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn committed:
!fixup address remaining comments, thanks!

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-compute-iv-end-values

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8955,24 +8854,79 @@ static void addCanonicalIVRecipes(VPlan &Plan, Type *IdxTy, bool HasNUW,
-/// Create resume phis in the scalar preheader for first-order recurrences and
-/// reductions and update the VPIRInstructions wrapping the original phis in the
-/// scalar header.
+/// Create and return a ResumePhi for \p WideIF, unless it is truncated. If the
```
fixed, thanks!
